rfc/reorder-waiting-executions.md

```diff
@@ -0,0 +1,116 @@
+
+
+## Timeline
+_The code for CLI command reorder implementation is almost ready for review. 
+Planning to implement this feature in two Phases._
+
+_Phase-1: completing CLI command implementation and make this feature available for customers to use and work on feedback if any._
+
+_Phase-2: Implementing GUI for reorder functionality by using the backend that is available from CLI reorder implementation in Phase-1._
+
+_Proposed timeline for Phase-1 would be 1-2 weeks as this feature is demoed for some Spinnaker users and got the feedback.
+Phase-2 would be 3-4 weeks as there are some existing issues in the waiting executions area of Spinnaker dashboard that needs to be fixed._
+
+## Design
+
+_The Re-ordering waiting executions of a pipeline can be done through Spin CLI command and Spin dashboard._
+
+_Introducing a new CLI command for reorder functionality of a waiting executions, the command will be used to Re-order waiting execution to UP/DOWN or Top/Bottom based on the current position of the provided execution._
+
+_Example Spin CLI Command_
+ >>spin pipeline execution reorder --execution-id 01GJQ601MC3SBACVNZB0XC5GHJ --reorder-action UP
+
+_New buttons will be added in Spinnaker dashboard(GUI) for each waiting execution to move the specific execution towards UP/Down or Top/Bottom based on the current position of the execution._
+
+_New Gate API endpoint will be created for external access_
+>>:8084/pipelines/<execution_id>/reorder?reorderAction=<UP/Down/Top/Bottom>
+
+_New Orca API endpoint will be created to access the same way_
+>>:8083/pipelines/<execution_id>/reorder?reorderAction=<UP/Down/Top/Bottom>
+
+_The main business logic of the feature lies in Spin Orca service, where the **CompoundExecutionOperator** will update the Redis repo using the existing class **RedisExecutionRepository** and update the pending executions to the user selected position in the Redis repo using the Class **RedisPendingExecutionService**._
+
+_A new handler class **ReorderWaitingExecutionHandler** will be created in orca-queue handler package, to handle the reorder of a waiting execution._
+
+**_Note_**
+
+_The Re-ordering will not be done for the executions which are not in Waiting state, an appropriate error message should be displayed._
+
+_In the Spinnaker dashboard(GUI), only the waiting executions should have buttons to reorder, buttons should be removed once the execution is started._
+
+
+### Dependencies
+
+_This proposal depends on existing services Spin CLI, Spin Deck, Spin Gate and Spin Orca._
+
+_No new dependencies are needed, existing systems will be sufficient to implement this feature_
+
+## Drawbacks
+
+_Currently there is no Re-order functionality of an execution_
+_Why should we **not** do this?_
```

Comment:
Some things that OFTEN hit:
* How do you know what the order SHOULD be - by doing it in FIFO you handle SHA's which may not have a particular order, but you REALLY want the latest kinda thing.
* E.g. Git triggers - it's possible that re-ordering would cause incorrect sha pulls on artifacts.  SO there IS a question in my mind of whether this should by default be visible?  It seems... rather specific to a very specific set of cases.  Docker triggers would be similar where unintentionally you could deploy older docker images.

Reply:
The pipeline design will have significant bearing on usage of triggers in the delivery workflow. If the Git commit id from the trigger is used in the pipeline and not head of the branch and uses the state of pipeline context then it is possible to have two pipeline executions deliver correctly even if order is changed. However, it makes sense that one would want the latest. 

There are couple of possibilities for reordering executions of the same pipeline
* The pipeline is used by triggers from multiple branches and build from main branch needs to be prioritized in the queue while others need to execute as well
* The pipeline is triggered by commits in same branch and multiple commits in the queue while the latest needs to be prioritized while others can be discarded

The first should use a different pipeline design and the second one should discard older executions in the queue as execution in different order can cause problems with identifying latest. This could then be an option of one execution at a time while discarding older execution in queue that has not yet started.

Reply:
Hi, 
Our Spinnaker users have similar request to re-order the waiting executions. But I think one of the most important reason is the FIFO is not working as expected in certain situations. I used to submit a ticket for this issue https://github.com/spinnaker/spinnaker/issues/6373. 

Can we consider to fix it first?

Reply:
Thanks @rjalander . Then it would be nice to also include this re-order feature . Our users will be happy to have it.

Reply:
One another scenario where Spinnaker users want to Re-order the waiting execution of same pipelines is,
- When the Pipeline has waiting executions from different source of triggers and If we want to prioritize the specific source of trigger to validate on priority for any business needs.

Reply:
DEFINITELY want the FIFO fix in first, then... more discussion on impacts/changes required for this, which aren't in my mind exactly "small" at the moment.  And this is MOSTLY tied to trigger impacts on executions and what data is pulled/used for triggers...